File: android/src/main/java/ucsd/ieeeqp/fa19/viewmodel/MmServiceViewModel.java
package ucsd.ieeeqp.fa19.viewmodel;

import android.app.Application;
import android.util.Log;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import call.MmProblemRequest;
import call.MmSolveStatus;
import kotlinx.coroutines.future.FutureKt;
import ucsd.ieeeqp.fa19.service.MmService;

import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.CompletableFuture;

import static androidx.constraintlayout.widget.Constraints.TAG;

public class MmServiceViewModel extends AndroidViewModel {
    public static final int NOT_LOGGED_IN = 1000, LOGIN_FAILED = 1001, LOGIN_SUCCESS = 1002;
    private static final long QUERY_PERIOD = 5000;

    private MmService mmService;
    private MutableLiveData<Integer> mmLoginStateLiveData = new MutableLiveData<>(NOT_LOGGED_IN);
    private MutableLiveData<List<MmSolveStatus>> mmSolveStatusLiveData = new MutableLiveData<>(new ArrayList<>());
    private Timer queryTimer;

    public MmServiceViewModel(@NonNull Application application) {
        super(application);
    }

    public void initService(String serverAuthCode) {
        invalidateService();
        mmService = new MmService(getApplication(), serverAuthCode);
    }

    public void invalidateService() {
        if (mmService != null) {
            mmService.close();
            mmService = null;
        }
    }

    public void loginToApi() {
        CompletableFuture<Boolean> loginFuture = FutureKt.asCompletableFuture(mmService.loginAsync());
        loginFuture.handleAsync((result, exception) -> {
            if (exception != null) {
                mmLoginStateLiveData.postValue(LOGIN_FAILED);
            } else if (result) {
                mmLoginStateLiveData.postValue(LOGIN_SUCCESS);
            } else {
                mmLoginStateLiveData.postValue(LOGIN_FAILED);
            }
            return null;
        });
    }

    public void startQueryAllStatuses() {
        if (queryTimer != null) {
            queryTimer.cancel();
        }
        queryTimer = new Timer();
        queryTimer.scheduleAtFixedRate(new TimerTask() {
            @Override
            public void run() {
                queryAllStatuses();
            }
        }, 0, QUERY_PERIOD);
    }

    private void queryAllStatuses() {
        CompletableFuture<List<MmSolveStatus>> future = FutureKt.asCompletableFuture(mmService.getAllProgressAsync());
        future.handleAsync((result, exception) -> {
            if (exception != null) {
                Log.d(TAG, "queryAllStatuses: result: " + result);
                mmSolveStatusLiveData.postValue(result);
            }
            return null;
        });
    }

    public void submitUnsolvedSchedule(MmProblemRequest request) {
        CompletableFuture<MmSolveStatus> future = FutureKt.asCompletableFuture(mmService.solveProblemAsync(request));
        future.handleAsync((result, exception) -> {
            String message = (exception == null) ? "Solve accepted; ID: " + result.getPid() : exception.getMessage();
            Toast.makeText(getApplication(), message, Toast.LENGTH_SHORT).show();
            return null;
        });
    }

    public LiveData<Integer> getMmLoginStateLiveData() {
        return mmLoginStateLiveData;
    }

    public LiveData<List<MmSolveStatus>> getMmSolveStatusLiveData() {
        return mmSolveStatusLiveData;
    }

    @Override
    protected void onCleared() {
        invalidateService();
        if (queryTimer != null) {
            queryTimer.cancel();
        }
    }
}

File: android/src/main/java/ucsd/ieeeqp/fa19/viewmodel/MmServiceViewModel.kt
package ucsd.ieeeqp.fa19.viewmodel

import android.app.Application
import android.util.Log
import android.widget.Toast
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import call.MmProblemRequest
import call.MmSolveStatus
import kotlinx.coroutines.launch
import ucsd.ieeeqp.fa19.service.MmService
import java.util.*
import kotlin.collections.ArrayList
import kotlin.concurrent.timerTask

class MmServiceViewModel(application: Application) : AndroidViewModel(application) {
    private val TAG: String = "MmServiceViewModel"
    private var mmService: MmService? = null
    private val mmLoginStateLiveData = MutableLiveData(NOT_LOGGED_IN)
    private val mmSolveStatusLiveData = MutableLiveData<ArrayList<MmSolveStatus>>(ArrayList())
    private val queryTimer: Timer = Timer()
    private var queryStatusTask: TimerTask? = null

    fun initService(serverAuthCode: String?) {
        invalidateService()
        mmService = MmService(getApplication(), serverAuthCode)
    }

    private fun invalidateService() {
        if (mmService != null) {
            mmService!!.close()
        }
    }

    fun loginToApi() = viewModelScope.launch {
        try {
            val loginSuccess = mmService!!.loginAsync().await()
            mmLoginStateLiveData.postValue(
                    if (loginSuccess) LOGIN_SUCCESS
                    else LOGIN_FAILED
            )
        } catch (e: Exception) {
            Log.e(TAG, "Login failed", e)
            mmLoginStateLiveData.postValue(LOGIN_FAILED)
        }
    }

    fun startQueryAllStatuses() {
        queryStatusTask?.cancel()
        queryStatusTask = timerTask {
            queryAllStatuses()
        }
        queryTimer.scheduleAtFixedRate(queryStatusTask, 0, QUERY_PERIOD)
    }

    private fun queryAllStatuses() = viewModelScope.launch {
        try {
            val allProgress = mmService!!.getAllProgressAsync().await()
            mmSolveStatusLiveData.apply {
                value!!.clear()
                value!!.addAll(allProgress)
                postValue(value)
            }
            Log.d(TAG, "queryAllStatuses: result: $allProgress")
        } catch (e: Exception) {
            Log.e(TAG, "queryAllStatuses: failed", e)
        }
    }

    fun submitUnsolvedSchedule(
            request: MmProblemRequest
    ) = viewModelScope.launch {
        try {
            val solveStatus = mmService!!.solveProblemAsync(request).await()
            val message = "Solve accepted; ID: ${solveStatus.pid}"
            Log.d(TAG, message)
            Toast.makeText(getApplication(), message, Toast.LENGTH_SHORT).show()
        } catch (e: Exception) {
            Log.e(TAG, "Submit unsolved schedule failed", e)
        }
        queryAllStatuses()
    }

    fun getMmLoginStateLiveData(): LiveData<Int> {
        return mmLoginStateLiveData
    }

    fun getMmSolveStatusLiveData(): LiveData<ArrayList<MmSolveStatus>> {
        return mmSolveStatusLiveData
    }

    override fun onCleared() {
        invalidateService()
        queryTimer.cancel()
    }

    companion object {
        const val NOT_LOGGED_IN = 1000
        const val LOGIN_FAILED = 1001
        const val LOGIN_SUCCESS = 1002
        private const val QUERY_PERIOD: Long = 5000
    }
}